internal/dag/builder_test.go

```diff
@@ -1071,6 +1071,124 @@ func TestDAGInsertGatewayAPI(t *testing.T) {
 				},
 			),
 		},
+		"insert basic single route with single header match, single path match": {
+			gateway: gatewayWithSelector,
+			objs: []interface{}{
+				kuardService,
+				&gatewayapi_v1alpha1.HTTPRoute{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "basic",
+						Namespace: "projectcontour",
+						Labels: map[string]string{
+							"app":  "contour",
+							"type": "controller",
+						},
+					},
+					Spec: gatewayapi_v1alpha1.HTTPRouteSpec{
+						Hostnames: []gatewayapi_v1alpha1.Hostname{
+							"test.projectcontour.io",
+						},
+						Rules: []gatewayapi_v1alpha1.HTTPRouteRule{{
+							Matches: []gatewayapi_v1alpha1.HTTPRouteMatch{{
+								Path: gatewayapi_v1alpha1.HTTPPathMatch{
+									Type:  "Prefix",
+									Value: "/",
+								},
+							}, {
+								Headers: &gatewayapi_v1alpha1.HTTPHeaderMatch{
+									Type:   "Exact",
+									Values: map[string]string{"foo": "bar"},
+								},
+							}},
+							ForwardTo: []gatewayapi_v1alpha1.HTTPRouteForwardTo{{
+								ServiceName: pointer.StringPtr("kuard"),
+								Port:        gatewayPort(8080),
+							}},
+						}},
+					},
+				},
+			},
+			want: listeners(
+				&Listener{
+					Port: 80,
+					VirtualHosts: virtualhosts(virtualhost("test.projectcontour.io", &Route{
+						PathMatchCondition: prefixString("/"),
+						HeaderMatchConditions: []HeaderMatchCondition{
+							{Name: "foo", Value: "bar", MatchType: "exact", Invert: false},
+						},
```

Comment:
Is this correct? From the documentation it seems this test case should have generated two separate routes, one that matches the header `/foo` against value `bar` *and* the path `/` and another route that matches only the path `/`

Reply:
To get the output this test is currently written with you would need just this:

```
						Rules: []gatewayapi_v1alpha1.HTTPRouteRule{{
							Matches: []gatewayapi_v1alpha1.HTTPRouteMatch{{
								Headers: &gatewayapi_v1alpha1.HTTPHeaderMatch{
									Type:   "Exact",
									Values: map[string]string{"foo": "bar"},
								},
							}},
							ForwardTo: ...
```

Since matches are only ANDed together within a `HTTPRouteMatch` list element, otherwise they are treated as separate route matches

See the description here: https://gateway-api.sigs.k8s.io/spec/#networking.x-k8s.io/v1alpha1.HTTPRouteRule

```
For example, take the following matches configuration:

matches:
- path:
value: "/foo"
headers:
values:
version: "2"
- path:
value: "/v2/foo"

For a request to match against this rule, a request should satisfy EITHER of the two conditions:
- path prefixed with /foo AND contains the header version: "2"
- path prefix of /v2/foo
```

Reply:
Updated this as weell.